crypto/hpke/hpke.c

```diff
@@ -0,0 +1,1370 @@
+    ctx = OPENSSL_zalloc(sizeof(OSSL_HPKE_CTX));
+    if (ctx == NULL)
+        return ctx;
+    ctx->libctx = libctx;
+    if (propq != NULL) {
+        ctx->propq = OPENSSL_strdup(propq);
+        if (ctx->propq == NULL)
+            goto err;
+    }
+    ctx->mode = mode;
+    ctx->suite = suite;
+    return ctx;
+
+err:
+    OSSL_HPKE_CTX_free(ctx);
+    return NULL;
+}
+
+void OSSL_HPKE_CTX_free(OSSL_HPKE_CTX *ctx)
+{
+    if (ctx == NULL)
+        return;
+    OPENSSL_free(ctx->propq);
+    OPENSSL_clear_free(ctx->exportersec, ctx->exporterseclen);
+    OPENSSL_free(ctx->pskid);
+    OPENSSL_clear_free(ctx->psk, ctx->psklen);
+    OPENSSL_clear_free(ctx->key, ctx->keylen);
+    OPENSSL_clear_free(ctx->nonce, ctx->noncelen);
+    OPENSSL_clear_free(ctx->shared_secret, ctx->shared_secretlen);
+    OPENSSL_clear_free(ctx->ikme, ctx->ikmelen);
+    EVP_PKEY_free(ctx->authpriv);
+    OPENSSL_free(ctx->authpub);
+
+    OPENSSL_free(ctx);
+    return;
+}
+
+int OSSL_HPKE_CTX_set1_psk(OSSL_HPKE_CTX *ctx,
+                           const char *pskid,
+                           const unsigned char *psk, size_t psklen)
+{
+    if (ctx == NULL || pskid == NULL || psk == NULL || psklen == 0) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if (ctx->mode != OSSL_HPKE_MODE_PSK
+        && ctx->mode != OSSL_HPKE_MODE_PSKAUTH) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    /* free previous value if any */
+    OPENSSL_free(ctx->pskid);
+    OPENSSL_clear_free(ctx->psk, ctx->psklen);
+    ctx->pskid = OPENSSL_strdup(pskid);
+    if (ctx->pskid == NULL)
+        goto err;
+    ctx->psk = OPENSSL_malloc(psklen);
+    if (ctx->psk == NULL)
+        goto err;
+    memcpy(ctx->psk, psk, psklen);
+    ctx->psklen = psklen;
+    return 1;
+err:
+    /* zap any new or old psk */
+    OPENSSL_free(ctx->pskid);
+    OPENSSL_clear_free(ctx->psk, ctx->psklen);
+    ctx->psklen = 0;
+    return 0;
+}
+
+int OSSL_HPKE_CTX_set1_ikme(OSSL_HPKE_CTX *ctx,
+                            const unsigned char *ikme, size_t ikmelen)
+{
+    if (ctx == NULL || ikme == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    OPENSSL_clear_free(ctx->ikme, ctx->ikmelen);
+    ctx->ikme = OPENSSL_malloc(ikmelen);
+    if (ctx->ikme == NULL)
+        return 0;
+    memcpy(ctx->ikme, ikme, ikmelen);
+    ctx->ikmelen = ikmelen;
+    return 1;
+}
+
+int OSSL_HPKE_CTX_set1_authpriv(OSSL_HPKE_CTX *ctx, EVP_PKEY *priv)
+{
+    if (ctx == NULL || priv == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if (ctx->mode != OSSL_HPKE_MODE_AUTH
+        && ctx->mode != OSSL_HPKE_MODE_PSKAUTH) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if (ctx->authpriv != NULL)
+        EVP_PKEY_free(ctx->authpriv);
+    ctx->authpriv = EVP_PKEY_dup(priv);
+    if (ctx->authpriv == NULL)
```

Comment:
return (ctx->authpriv != NULL);

Reply:
Hmm, I think it's clearer as-is and no benefit from changing that way I can see. Happy to change if there's one.